### ISSUE TITLE
Nextjs dev compiles each page at request time... why?

### ISSUE THREAD

**Issue by ajhool** [zeit/next.js#6554]
# Bug report

## Describe the bug

In dev mode, Next.js compiles each web page at request time. I can't imagine what the reason for this is? This leads to the UI bug of clicking a next/link and waiting 5+ seconds to see a change on the page -- while that might be a "feature" it really feels like a bug during development and the `next dev` experience feels very buggy. Why not compile pages at file-change time so that page navigation feels normal?

## To Reproduce

Create a next app and execute `next dev`

## Expected behavior

Change a file while a local next app is running -> next rebuilds -> clicking a next/link or navigating to a page does not take 5-10 seconds on the first click and it feels like a useable website.

If somebody finds this feature useful, then a configuration option to disable it would be great.

## System information
 - OS: macOS
 - Browser (if applies) [e.g. chrome, safari]
 - Version of Next.js: 8.0.0

**Comment by Timer** [comment]
This is done intentionally to minimize memory and CPU usage, and to keep rebuilds fast!
We're sorry you find the experience less than optimal, but this will likely remain as-is for the time being.

We have some things in the pipeline that should make initial compilation faster to alleviate this! 🏎 

Thanks again for bringing this to our attention.

**Comment by bz-stripe** [comment]
Are there any debug suggestions if rebuilds are very slow? On the order for 2-3s per page transition?

**Comment by timneutkens** [comment]
Hey @bz-stripe could you send me a message on twitter.com/timneutkens, I'd love to learn more about your specific use case.

**Comment by borekb** [comment]
If I have enough CPU / memory, it would nice if `next dev` had a flag or something to pre-compile everything in the background.

- `next start`: slow to build, fast to browse
- `next dev`: (relatively) fast to build, slow to browse
- `next dev --precompile`: (relatively) fast to build, fast to browse after a few seconds

Now sure how this would work on huge sites but in our case (relatively complex component tree but relatively few entrypoints), this would improve the DX greatly.

**Comment by ajhool** [comment]
Just to update this issue, I had posted this issue before Next added a "loading" symbol while the page was loading. So, you would click a button to go to the next page and nothing would happen for ~5 seconds. The loading symbol has helped significantly although I do agree that I'd rather have a slow build and fast usage than a fast build and slow usage, so @borekb 's flag suggestion is a good idea IMIO.

**Comment by durchanek** [comment]
Another use case is TypeScript, where precompiling pages would show possible typing issues instantly instead of when opening the page for the first time.

**Comment by imgntn** [comment]
the compile-time lag when switching pages makes it pretty difficult to work on the UX for anything that transitions across page boundaries :/

**Comment by salazarm** [comment]
As a workaround if you know you're working across multiple pages you could just import those pages into page you load first during development.

**Comment by ajhool** [comment]
@Timer Would you please consider reopening this issue based on the feedback here? There are some good ideas about setting a compile-flag.

I understand the idea of having fast-compile, but it's reallyyyyyy wonky to not be able to locally test page transitions.

**Comment by mifrej** [comment]
Any update on this. It's an issue for testing in Cypress e2e library. Each page change triggers building page with different timeouts. Quite indeterministic

**Comment by sanojsilva** [comment]
> Any update on this. It's an issue for testing in Cypress e2e library. Each page change triggers building page with different timeouts. Quite indeterministic

Yes same problem

**Comment by markschellhas** [comment]
@Timer this issue is actually using much more CPU than necessary. For anyone having these issues, updating to the latest version of Next.js, and using as few server rendered files as possible might help.

Or just switch to Sapper and Svelte... you'll thank me. I love Next.js since very early on, it's sloooowed down my development time to unacceptable levels.

**Comment by cristian-eriomenco** [comment]
What is the reason for this ticket to be closed?

**Comment by samipshah100** [comment]
Nextjs can compile the rest of the pages after loading the first page. This is slowing my dev time a lot. Please do consider these requests by your users if possible. Thanks for the cool product :)

**Comment by ethanwilkins** [comment]
Why was this closed?

**Comment by imgntn** [comment]
> Why was this closed?

not spam

**Comment by eric-burel** [comment]
I think everyone agrees that this is a pain point, especially when your number of pages is small (SaaS applications for instance). But the question is maybe how can we implement the `--precompile` flag proposed by @borekb ? What are the technical difficulties?

- Is there a palliative already? Like importing the other pages dynamically in dev mode to trigger their build in the background?
- In which order should the background pages be prebuilt? Should there be an order?
- Shouldn't we focus on per-page build time instead? I also encounter those "long" build times (a few seconds of build time on a powerful machine), using TypeScript and ton of library, and never managed to reduce it. Maybe we lack the tools and patterns to fasten build/debug slow builds?
- When should background pages be prebuilt to avoid high CPU consumption? Is it realistic to check CPU usage for instance and run the prebuild when it is low ? Or maybe a few ms after the current page has been built ?

**Comment by visnup** [comment]
A few thoughts about workarounds (none great):

If a page link is very likely to be loaded after another, you could use [prefetch](https://developer.mozilla.org/en-US/docs/Glossary/Prefetch#prefetch_attribute_value) tags in the header of the source page to have browsers start fetching the related pages.

@mifrej @sanojsilva For Cypress in CI, I've switched to running `next build` and using `next start` since within CI there's no chance that source code changes happen. This also has the added benefit that what my Cypress tests are running against are closer to a production build. Running Cypress locally, I still use `next dev` though, but I also have a suite called "_wait.js" that visits my main pages so that they're ready by the time the other tests visit them. Combining the prefetch idea above, I could see having an "index" page that included prefetches to all the others which Cypress could load as a first step.

**Comment by bz-stripe** [comment]
Does the prefetch tag actually trigger the dev server to build the next bundle? Or is that a suggestion for a future feature?

**Comment by drewlustro** [comment]
@bz-stripe - It's a suggestion we're hoping gets traction with Next.js maintainers.

**Comment by HarryF514** [comment]
+1 need this.

**Comment by elado** [comment]
@kyle-ssg 
> Need this for automated tests
I use https://www.npmjs.com/package/start-server-and-test and it works great.

**Comment by elado** [comment]
Also wanted something like that. Many times I `yarn dev` but go do something else while I wait, then when I access localhost I wait more for the home route to compile.


I ended up adding `wait-on` in parallel to starting the server:

```json
    "dev": "npm-run-all --parallel dev-server precompile-routes",
    "dev-server": "next dev",
    "precompile-routes": "wait-on http-get://localhost:3000",
```

requires `wait-on` devDep.

**Comment by rnbrady** [comment]
In my Typescript projects this causes debugging breakpoints to be unbound until the first use of the api page in question.

**Comment by eric-burel** [comment]
I seem to experience a slightly related issue: when manually clicking on links within my app,  I see requests to `webpack-hmr?page=foobar` => this seems to unexpectly trigger page reload, thus breaking the resetting the app state, instead of an SPA page transition. I never reproduced in development but I do in cypress test.

Any idea how I could strictly disable any page reload in this scenario, and yet keep Next running in dev mode?

**Comment by rtrembecky** [comment]
running `yarn dev` I'm currently experiencing the following issue:
1. I set some value in react context
2. I navigate using `router.push()`
3. the page compiles
4. I access the value in the context - it's empty.

This is not the case when I visit and let the page compile beforehand. As I have more transitions like this, it renders the dev mode unusable for me.

**Comment by dankobgd** [comment]
so this is not a problem to you guys??
why is this closed? this behavior doesn't make sense.
you want my app to be faster by waiting 5 seconds every time, don't understand.
how can stupid gatsby be fast and you can't be fast.
what if i have a fast pc? i still need to wait forever
stupid feature

**Comment by CodeJuggernaut** [comment]
We really need a `--pre-compile` flag - or at least an easy way to build the app in non production-optimized mode.

**Comment by shady00999** [comment]
Was this resolved with the new release or why is the issue closed?

**Comment by rtrembecky** [comment]
Hi guys! I'm coming back with a big surprise on my face because it seems that indeed in Next.js 12 this is fixed! 🎉 
I was now able to go through the exact flow I mentioned above on `yarn dev` without losing context. I think this issue is finally rightfully closed 🙂 
Does anybody else care to verify as well?

**Comment by drewlustro** [comment]
Not certain, but I suspect it's still compiling...perhaps you've noticed
how fast `swc` is?

**Comment by rtrembecky** [comment]
True, I also switched to SWC with the next.js upgrade. I'll test which one does it for me.

**Comment by rtrembecky** [comment]
Ah, well, I'm also talking about a separate (but related) problem. It is true the pages are still compiled on request, but the problem in my case was losing context on such navigation. The context is not lost anymore even though the pages are still compiled on the run.

**Comment by elton2048** [comment]
The compile time is non-negligible when the project runs in docker env even using latest next.js

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
